### PR TITLE
Implemetation

### PR DESCRIPTION
**Pull request by ebachelet**
Hi guys

I have implemented herder_checker qc and test accordingly. Will I pass the severe Curtis criteria?

### CODE REVIEW

File: banzai/tests/test_stats.py [hunk 1]
```diff
@@ -529,15 +529,22 @@ def test_mad_3d_axis_1_mask_none():
...
         a = np.random.normal(mean, sigma, size=(size1, size2, size3))
+        b = a.copy()
+        
         expected = np.median(np.abs(a.astype(np.float32) - np.median(a.astype(np.float32), axis=2).reshape(size1, size2, 1)), axis=2)
-        actual = stats.median_absolute_deviation(a, axis=2)
-        np.testing.assert_allclose(actual, expected.astype(np.float32), atol=1e-4)
+        actual = stats.median_absolute_deviation(b, axis=2)
+        print(a[:,:,0])
+        print(np.median(a, axis=2)[0,0])
+        print(actual[0,0], expected[0,0])
```
Comment (cmccully) [inline banzai/tests/test_stats.py]: Tests should not print anything.

File: banzai/qc/header_checker.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    def do_stage(self, images):
+        """ Performs several checks on the header, see the relative functions.
+        :param list images: a list of image object.
+        :returns: the list of image object after check
+        :rtype: list
+        """
+        for image in images:
+
+
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Only skip one line between these blocks.

File: banzai/qc/header_checker.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            self.check_ra_range(image)
+            self.check_dec_range(image)
+            self.check_exptime_value(image)
+
+        return images
+
+    def check_header_keyword_present(self, keyword, image):
+        """ Return a warning if the keyword is not in the image header.
+        :param str keyword: the stry keyword you look inside the header
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: This format is for the docstring is fine, but I tend to try to follow the astropy convention which is slightly different.

File: banzai/qc/header_checker.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+"""
+This module performs controls on several critical parameters of image header, contains in
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: I would rephrase this to be something like "This module performs basic sanity checks that the main image header keywords are the correct format and validates their values."

File: banzai/qc/header_checker.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+header_expected_format dictionnary.
+@author:ebachelet
+"""
+from banzai.stages import Stage
+from banzai import logs
+
+class HeaderSanity(Stage):
+    """
+       This class contains the needed function/definitions to performs the header check.
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: This docstring does not really have any information in it. Maybe something like "Stage to validate important header keywords" or something like that.

File: banzai/qc/header_checker.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    """
+
+    def __init__(self, pipeline_context):
+        super(HeaderSanity, self).__init__(pipeline_context)
+
+        self.header_expected_format = {'RA': str, 'DEC': str, 'CAT-RA': str, 'CAT-DEC': str,
+                                       'OFST-RA': str, 'OFST-DEC': str, 'TPT-RA': str,
+                                       'TPT-DEC': str, 'PM-RA': str, 'PM-DEC': str,
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: I am nervous to include all of the variations of RA and Dec here because they may change over time. I don't really have a good solution to block against that though.

File: banzai/qc/header_checker.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                                       'CRVAL1': float, 'CRVAL2': float, 'CRPIX1': int,
+                                       'CRPIX2': int, 'EXPTIME': float}
+
+    @property
+    def group_by_keywords(self):
+        return None
+
+    def do_stage(self, images):
+        """ Performs several checks on the header, see the relative functions.
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Maybe: "Run stage to validate header"

File: banzai/qc/header_checker.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        :param list images: a list of image object.
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: list of banzais.images.Image

File: banzai/qc/header_checker.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        :returns: the list of image object after check
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: list of banzai.images.Image
List of validated images.

Do we ever reject images? If so we should mention that here.

File: banzai/qc/header_checker.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                self.check_header_na(keyword, image)
+
+            self.check_ra_range(image)
+            self.check_dec_range(image)
+            self.check_exptime_value(image)
+
+        return images
+
+    def check_header_keyword_present(self, keyword, image):
+        """ Return a warning if the keyword is not in the image header.
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Not return. Logs.

File: banzai/qc/header_checker.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        :param str keyword: the stry keyword you look inside the header
+        :param object image: an image object
+        """
+        header = image.header
+
+        logging_tags = logs.image_config_to_tags(image, self.group_by_keywords)
+
+        if keyword not in header:
+            sentence = 'WARNING : The header key ' + keyword + ' is not in image header!'
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Don't have warning in the message itself. Use logger.warn and that will produce the warning message correctly.

File: banzai/qc/header_checker.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+            self.logger.error(sentence, extra=logging_tags)
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Use warning if you want it to be a warning. Or error if you want it to be an error.

File: banzai/qc/header_checker.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+            return 'WARNING'
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: You do not need to return anything here.

File: banzai/qc/header_checker.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        logging_tags = logs.image_config_to_tags(image, self.group_by_keywords)
+
+        if not isinstance(header_value, self.header_expected_format[keyword]):
+            sentence = ('The header key ' + keyword + ' got an unexpected format : ' + type(
+                header_value).__name__ + ' in place of ' + self.header_expected_format[
+                    keyword].__name__)
+
+            self.logger.error(sentence, extra=logging_tags)
+
+        return
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: You do not need a return if you return None.

File: banzai/qc/header_checker.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        if keyword not in header:
+            sentence = 'WARNING : The header key ' + keyword + ' is not in image header!'
+
+            self.logger.error(sentence, extra=logging_tags)
+
+            return 'WARNING'
+        return
+
+    def check_header_format(self, keyword, image):
+        """ Return a warning if the keyword is not the expected type
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Again log a warning.

File: banzai/qc/header_checker.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            in the image header.
+        :param str keyword: the stry keyword you look inside the header
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: stry?

Comment (cmccully) [inline banzai/qc/header_checker.py]: Also, maybe just "header keyword to check"

File: banzai/qc/header_checker.py (hunk 8 above)
Comment (cmccully) [inline banzai/qc/header_checker.py]: Use this throughout.

File: banzai/qc/header_checker.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                header_value).__name__ + ' in place of ' + self.header_expected_format[
+                    keyword].__name__)
+
+            self.logger.error(sentence, extra=logging_tags)
+
+        return
+
+    def check_header_na(self, keyword, image):
+        """ Return a warning if the keyword is 'N/A' instead of the
+            expected type in the image header.
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: This is good, but all bias frames and dark frames have RA and Dec as N/A. So we don't want to produce warning for every single one of those calibration files.

File: banzai/qc/header_checker.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        :param str keyword: the stry keyword you look inside the header
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: See comment above.

File: banzai/qc/header_checker.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        if isinstance(header_value, str):
+
+            if 'N/A' in header_value:
+                sentence = 'The header key ' + keyword + ' got the unexpected value : N/A'
+                self.logger.error(sentence, extra=logging_tags)
+
+        return
+
+    def check_ra_range(self, image):
+        """ Return an error if the keyword right_ascension is not inside
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Again this doesn't return an error. See below as well. This seems like some copy and pasted code. Could you abstract this a bit into a single function?

File: banzai/tests/test_header_checker.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    ny = 103
+
+    images = [FakeImage(nx=nx, ny=ny) for x in range(6)]
+
+    for image in images:
+
+        tester.check_header_keyword_present('EXPTIME', image)
+        assert tester.logger.error.called
+
+    assert len(images) == 6
```
Comment (cmccully) [inline banzai/tests/test_header_checker.py]: This would be where we could check if images are being rejected correctly.

File: banzai/tests/test_header_checker.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    tester.logger.error = mock.MagicMock()
+    nx = 101
+    ny = 103
+
+    images = [FakeImage(nx=nx, ny=ny) for x in range(6)]
+    for image in images:
+        image.header['RA'] = 20
+
+        tester.check_header_format('RA', image)
+        assert tester.logger.error.called
```
Comment (cmccully) [inline banzai/tests/test_header_checker.py]: This makes me nervous a bit. What if we decide to switch RA and Dec to decimal instead of segsidecimal?

File: banzai/tests/test_stats.py (hunk 1 above)
Comment (ebachelet) [inline banzai/tests/test_stats.py]: OK done, this comes from an old discussion we had on low number statistics

File: banzai/qc/header_checker.py (hunk 9 above)
Comment (ebachelet) [inline banzai/qc/header_checker.py]: I think we decided a while ago to not reject anything from this module, because some info could be found again easily, examples date of observation or RA DEC

File: banzai/qc/header_checker.py (hunk 17 above)
Comment (ebachelet) [inline banzai/qc/header_checker.py]: Does these BIAS and DARK have specificity inside OBSTYPE for example? If yes, I can put an exception.

File: banzai/tests/test_header_checker.py (hunk 21 above)
Comment (ebachelet) [inline banzai/tests/test_header_checker.py]: See my first comment.

File: banzai/tests/test_header_checker.py (hunk 20 above)
Comment (ebachelet) [inline banzai/tests/test_header_checker.py]: We decided to not reject frame form this module. Just to indicates logs. We can change this if needed now, but this needs discussions.

File: banzai/qc/header_checker.py [hunk 22]
```diff
@@ -58,15 +71,19 @@ def check_header_keyword_present(self, keyword, image):
 
             self.logger.error(sentence, extra=logging_tags)
```
Comment (cmccully) [inline banzai/qc/header_checker.py]: Is this supposed to to log an error or warning?
